sound/soc/sof/ipc4-topology.c

```diff
@@ -173,6 +173,40 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
 	}
 }
 
+static int sof_ipc4_setup_process_set_cdata(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct sof_ipc4_process *process = swidget->private;
+	struct sof_ipc4_control_data *control_data;
+	const struct snd_kcontrol_new *kc;
+	struct snd_sof_control *scontrol;
+	struct soc_bytes_ext *sbe;
+	int i;
+
+	/*
+	 * There should be zero or one byte kcontrol for module instance
+	 * initialization per process module, if more than one are provided
+	 * from topology, the first found in widget kcontrol list will be used.
+	 */
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+
+		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
+			continue;
+		sbe = (struct soc_bytes_ext *)kc->private_value;
+		scontrol = sbe->dobj.private;
```

Comment:
In `sof_control_load()`, we assign `sbe->dobj.private = scontrol` for byte/enum/mixer control, in `sof_ipc4_control_load_bytes()` we have checked the `SOF_IPC4_ABI_MAGIC` for all byte controls, derefence is safe here. The EQ coefficients or WoV modules is either set by module init IPC or set_large_config IPC, they will have corresponding byte control with type MOD_INIT/SET_LARGE_CONFIG

Reply:
@aiChaoSONG, can you clarify this? Are you saying that the EQ parameter for the WoV module is either MOD_INIT_INSTANCE type _or_ LARGE_CONFIG? Is this correct?
I'm asking it because you can not have both at the same time for the very same payload. Yes, you can have separate controls for the MOD_INIT_INSTANCE blob and for the LARGE_CONFIG type but they are separate and they are not sharing any information, meaning you need separate binary for each.

Reply:
It is also not clear what to do with non LARGE_CONFIG (and MOD_CONFIG) type with the user space kcontrol code. My current implementation just allows updating the blob for MOD_INIT_INSTANCE and that's it.

Reply:
> I'm asking it because you can not have both at the same time for the very same payload. Yes, you can have separate controls for the MOD_INIT_INSTANCE blob and for the LARGE_CONFIG type but they are separate and they are not sharing any information, meaning you need separate binary for each.

@ujfalusi why do you even need separate controls. For example, the blob in the topology can set the blob type to MOD_INIT_INSTANCE so that the EQ coeff will be sent to the module during init. But using the same kcontrol, a user can generate a new blob with LARGE_CONFIG type to update the coeff no?

Reply:
> In `sof_control_load()`, we assign `sbe->dobj.private = scontrol` for byte/enum/mixer control, in `sof_ipc4_control_load_bytes()` we have checked the `SOF_IPC4_ABI_MAGIC` for all byte controls, derefence is safe here. The EQ coefficients or WoV modules is either set by module init IPC or set_large_config IPC, they will have corresponding byte control with type MOD_INIT/SET_LARGE_CONFIG

It doesn't answer to my question. You are looking on the kcontrols exposed by a widget, I don't see how you can differentiate between the ones used for FUNCTIONAL stuff like actual coefficients and whatnot, and the ones used for IPC4 PLUMBING such a module initalization.

Reply:
@plbossart the whole point of introducing the struct sof_ipc4_control_data and the IPC4 ABI magic in one of the previous PRs was to do this distinction right? Can I please elaborate on why this isnt enough?

Reply:
> > I'm asking it because you can not have both at the same time for the very same payload. Yes, you can have separate controls for the MOD_INIT_INSTANCE blob and for the LARGE_CONFIG type but they are separate and they are not sharing any information, meaning you need separate binary for each.
> 
> @ujfalusi why do you even need separate controls. For example, the blob in the topology can set the blob type to MOD_INIT_INSTANCE so that the EQ coeff will be sent to the module during init. But using the same kcontrol, a user can generate a new blob with LARGE_CONFIG type to update the coeff no?

I don't think this would ever work, the blob which we stored is a 'MOD_INIT_INSTANCE' type and it is likely not compatible with 'LARGE_CONFIG' or 'MOD_CONFIG', needs some surgery on the blob to "convert" it and this is just not going to be a sane thing in the kernel. Not to mention that some modules might only support 'MOD_INIT_INSTANCE' and nothing else.
You need flags, offsets, cutlines, re-packing envelops here and there to be able to do that.